packages/compiler-cli/src/ngtsc/core/api/src/interfaces.ts

```diff
@@ -49,6 +49,52 @@ export interface ResourceHost {
    * or `undefined` if this is not an incremental build.
    */
   getModifiedResourceFiles?(): Set<string>|undefined;
+
+  /**
+   * Transform an inline or external resource asynchronously.
+   * It is assumed the consumer of the corresponding `Program` will call
+   * `loadNgStructureAsync()`. Using outside `loadNgStructureAsync()` will
+   * cause a diagnostics error or an exception to be thrown.
+   * Only style resources are currently supported.
+   *
+   * @param data The resource data to transform.
+   * @param context Information regarding the resource such as the type and containing file.
+   * @returns A promise of either the transformed resource data or null if no transformation occurs.
+   */
+  transformResource?
+      (data: string, context: ResourceHostContext): Promise<TransformResourceResult|null>;
+}
+
+/**
+ * Contextual information used by members of the ResourceHost interface.
+ */
+export interface ResourceHostContext {
+  /**
+   * The type of the component resource. Templates are not yet supported.
+   * * Resources referenced via a component's `styles` or `styleUrls` properties are of
+   * type `style`.
+   */
+  readonly type: 'style';
+  /**
+   * The absolute path to the resource file. If the resource is inline, the value will be undefined.
+   */
+  readonly resourceFile?: string;
```

Comment:
We've generally been avoiding optional properties in the compiler:

```suggestion
  readonly resourceFile: string|null;
```

Reply:
Outside the DOM API, `null` is rarely used.  Any particular reason for avoiding optional properties?

Reply:
The primary reason for required properties is that they can't be forgotten about when creating such objects. As for `null`, that's us being explicit about an absent value. I don't feel too strongly about it, though.